Codechange: Replace path cache deques with vectors.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

RoadVehicle path cache has two std::deques which is quite awkward to manage, and also quite memory hungry.
Ship path cache use a single std::deque.
std::deque was used to be able to push/pop from either end.

#12275 attempts to partially address RoadVehicle's path cache by hiding the fact it has two deques.

With this PR I address both the split, and the memory usage.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

Instead, change both to use a single `std::vector` each. For `RoadVehicle`s this now contains a `std::pair<Trackdir, TileIndex>` which simplifies usage.

This is now push/popped from the back, so sometimes needs to be reversed.

Also a special-case for ships that was copied across in the original implementation, which no longer exists for ships, has been removed for RVs. As far as I can tell this never made any sense, as an RV destination tile cannot have a path finding choice, so would never appear in the path cache anyway.

| Type | Master | PR |
| ---- | ------ | -- |
| Ship | 640 bytes + | 584 bytes + |
| RoadVehicle | 776 bytes + | 640 bytes + |

+ Extra memory outside the object is used by both deque and vector.

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

Performance not compared. No regression expected for Ships, but RVs has a little bit of erasing which could perhaps be improved.

~~Saveload code for RoadVehicle is a bit gnarly as std::pair is difficult to saveload, so it always uses two temporary vectors. Maybe using a struct instead would help.~~

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

Kuhnovic commented on src/saveload/vehicle_sl.cpp:
```diff
@@ -841,22 +845,56 @@ class SlVehicleRoadVeh : public DefaultSaveLoadHandler<SlVehicleRoadVeh, Vehicle
 	void Save(Vehicle *v) const override
 	{
 		if (v->type != VEH_ROAD) return;
+		ClearPathCache();
```
I think it makes more sense to call ClearPathCache from JoinPathCache / SplitPathCache directly, since only those two functions deal with the static vectors.

PeterN replied:
The order of operations is different:

Save is Clear → Split → SlObject.
Load is Clear → SlObject → Join

Kuhnovic replied:
Ah yeah, then ignore this :)

Kuhnovic commented on src/saveload/vehicle_sl.cpp:
```diff
@@ -841,22 +845,56 @@ class SlVehicleRoadVeh : public DefaultSaveLoadHandler<SlVehicleRoadVeh, Vehicle
+		}
+	}
+
 	void Save(Vehicle *v) const override
```
A reference probably makes more sense in this context, a nullptr vehicle doesn't make much sense. But I guess I lot of the code here is C-style passing pointers so in that sense it would be consistent. Whatever you prefer.

Same for the other functions btw.

PeterN replied:
It's the interface to SaveLoad that uses pointers, I'm not adding to changing anything here.

Kuhnovic commented on src/saveload/vehicle_sl.cpp:
```diff
@@ -841,22 +845,56 @@ class SlVehicleRoadVeh : public DefaultSaveLoadHandler<SlVehicleRoadVeh, Vehicle
+		}
+	}
+
+	void JoinPathCache(RoadVehicle *rv) const
```
Isn't "merge"a more correct them instead of "join"? Join immediately makes me thing about (string) concatenation, where merging is more like "combine" or "mix".

Not a native English speaker, just my 2ct

PeterN replied:
Fair :)

PeterN committed:
Codechange: Replace path cache queues with vectors.

Ship and RoadVehicle path caches use a std::deque, which is quite memory hungry, especially for RoadVehicle which has two.
std::deque was used to be able to push/pop from either end.

Change to use a single std::vector each, which is now push/popped from the back.

PeterN committed:
Codefix: Clear and reserve storage when loading vectors from savegames.

LordAro commented on src/roadveh.h:
```diff
@@ -81,25 +81,16 @@ static const uint8_t RV_OVERTAKE_TIMEOUT = 35;
-struct RoadVehPathCache {
-	std::deque<Trackdir> td;
-	std::deque<TileIndex> tile;
+struct RoadVehPath {
```
maybe `RoadVehPathNode` ? `RoadVehPath` suggests it's the entire path

also for all the other structs

(also, doxygen?)

Kuhnovic replied:
Node seems to suggest it's a pathfinder node, which it isn't. I would just call it RoadVehPathElement or RoadVehPathEntry, something along those lines
